runtime/src/accounts_db.rs

```diff
@@ -4918,6 +4944,38 @@ impl AccountsDb {
         self.do_load(ancestors, pubkey, None, load_hint, LoadZeroLamports::None)
     }
 
+    pub fn account_matches_owners(
+        &self,
+        ancestors: &Ancestors,
+        account: &Pubkey,
+        owners: &[&Pubkey],
+    ) -> Result<(), MatchAccountOwnerError> {
+        let (slot, storage_location, _maybe_account_accesor) = self
+            .read_index_for_accessor_or_load_slow(ancestors, account, None, false)
+            .ok_or(MatchAccountOwnerError::UnableToLoad)?;
+
+        if !storage_location.is_cached() {
+            let result = self.read_only_accounts_cache.load(*account, slot);
+            if let Some(account) = result {
+                return (!account.is_zero_lamport() && owners.contains(&account.owner()))
+                    .then_some(())
+                    .ok_or(MatchAccountOwnerError::NoMatch);
+            }
+        }
+
+        let (account_accessor, _slot) = self
+            .retry_to_get_account_accessor(
```

Comment:
it doesn't look like zero lamport is handled for the case where we load from the append vec. I think you need this same type of code here:
```
!account.is_zero_lamport() && owners.contains(&account.owner()))
```

Reply:
The `append_vec.rs` makes the check and returns the error for 0 lamports.
The `append_vec` API is not returning the account meta, otherwise we could've consolidated this check in `accounts_db.rs`.

Reply:
https://github.com/solana-labs/solana/pull/30154/files#diff-f369d6a7bff2bb6f2bbbd60cdb9f8255b51a60a7af47b135a9fa592178266294R626

Reply:
my apologies. I was looking at the CURRENT master code to see that we weren't doing anything special on zero lamport accounts...